Panic on LogstreamerInput shutdown

```
2015/05/14 12:42:26 Input started: LogstreamerInput
^C2015/05/14 12:42:27 Shutdown initiated.
2015/05/14 12:42:27 Stop message sent to input 'LogstreamerInput'
2015/05/14 12:42:27 Waiting for decoders shutdown
2015/05/14 12:42:27 Decoder 'LogstreamerInput-ProtobufDecoder-3': stopped
2015/05/14 12:42:27 Stop message sent to decoder 'LogstreamerInput-ProtobufDecoder-2'
panic: close of closed channel

goroutine 36 [running]:
github.com/mozilla-services/heka/pipeline.(*PipelineConfig).StopDecoderRunner(0xc20807e1c0, 0x7ff488933498, 0xc214238080, 0xc214226000)
	/work/git/data-pipeline/build/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/config.go:324 +0x188
github.com/mozilla-services/heka/pipeline.(*deliverer).Done(0xc20800b620)
	/work/git/data-pipeline/build/heka/build/heka/src/github.com/mozilla-services/heka/pipeline/plugin_runners.go:138 +0x51
github.com/mozilla-services/heka/plugins/logstreamer.(*LogstreamInput).Run(0xc208064300, 0x7ff48892df68, 0xc208001c20, 0x7ff488933300, 0xc20807e1c0, 0xc2142260c0, 0x7ff4889335c0, 0xc20800b620, 0x7ff4889335f8, 0xc21423e0b0)
	/work/git/data-pipeline/build/heka/build/heka/src/github.com/mozilla-services/heka/plugins/logstreamer/logstreamer_input.go:331 +0x328
created by github.com/mozilla-services/heka/plugins/logstreamer.(*LogstreamerInput).startLogstreamInput
	/work/git/data-pipeline/build/heka/build/heka/src/github.com/mozilla-services/heka/plugins/logstreamer/logstreamer_input.go:186 +0x291
```
Configuration

```toml
[hekad]
maxprocs = 4
base_dir = "."
share_dir = "."
sample_denominator = 1000000
max_message_size = 1000000


[LogstreamerInput]
log_directory = "/testdir//input"
file_match = '(?P<Name>\w+)/(?P<Seq>\d*)\.log'
priority = ["Seq"]
differentiator = ["Name"]
decoder = "ProtobufDecoder"
```

To reproduce startup Heka and shut it down immediately.

--- a/CHANGES.txt
+++ b/CHANGES.txt
@@ -9,6 +9,9 @@ Bug Handling
 * Removed state from PluginMaker to remove race conditions during plugin
   construction (#1532).
 
+* Get decoder lock before cleaning up decoders during pipeline shutdown to
+  avoid race condition panics during exit (#1531).
+
 0.9.2 (2015-04-22)
 ==================
 

--- a/pipeline/pipeline_runner.go
+++ b/pipeline/pipeline_runner.go
@@ -4,7 +4,7 @@
 # You can obtain one at http://mozilla.org/MPL/2.0/.
 #
 # The Initial Developer of the Original Code is the Mozilla Foundation.
-# Portions created by the Initial Developer are Copyright (C) 2012-2014
+# Portions created by the Initial Developer are Copyright (C) 2012-2015
 # the Initial Developer. All Rights Reserved.
 #
 # Contributor(s):
@@ -17,8 +17,6 @@
 package pipeline
 
 import (
-	"github.com/mozilla-services/heka/message"
-	"github.com/rafrombrc/go-notify"
 	"os"
 	"os/signal"
 	"path/filepath"
@@ -27,6 +25,9 @@ import (
 	"sync/atomic"
 	"syscall"
 	"time"
+
+	"github.com/mozilla-services/heka/message"
+	notify "github.com/rafrombrc/go-notify"
 )
 
 const (
@@ -302,11 +303,14 @@ func Run(config *PipelineConfig) {
 	config.inputsLock.Unlock()
 	config.inputsWg.Wait()
 
+	config.allDecodersLock.Lock()
 	LogInfo.Println("Waiting for decoders shutdown")
 	for _, decoder := range config.allDecoders {
 		close(decoder.InChan())
-		LogError.Printf("Stop message sent to decoder '%s'", decoder.Name())
+		LogInfo.Printf("Stop message sent to decoder '%s'", decoder.Name())
 	}
+	config.allDecoders = config.allDecoders[:0]
+	config.allDecodersLock.Unlock()
 	config.decodersWg.Wait()
 	LogInfo.Println("Decoders shutdown complete")
 

--- a/pipeline/plugin_maker.go
+++ b/pipeline/plugin_maker.go
@@ -317,7 +317,6 @@ func (m *pluginMaker) makeSplitterRunner(name string, config interface{}, splitt
 func (m *pluginMaker) makeInputRunner(name string, config interface{}, input Input,
 	defaultTick uint) (InputRunner, error) {
 
-	fmt.Println("Before fetching")
 	commonConfig, err := m.prepCommonTypedConfig()
 	if err != nil {
 		return nil, fmt.Errorf("Can't prep common typed config: %s", err.Error())

